library

pazz opened the pull request:
This refactors shepherd into a library crate with an additional `main.rs` module for the command-line tool.

I do this because the added modularity allows for easier re-use. I want in particular another cli tool that eats the same input and generates prism models for later comparison.

the "good" news is that this split already found some "bugs": Only on library crates, `cargo test` will run documentation tests and some of the code examples in documentation strings are actually broken.

@Numero7 do you want to fix or remove these?

Numero7 commented:
@pazz it would be cool to fix those documentation tests, please tell me in case want me to do it

pazz commented:
They seem to pass now, but you should go over this at some point:

What is the "max_finite_coordinate" argument in [`DownSet.safe_pre_image`](https://github.com/SynthesisLab/shepherd/blob/main/src/downset.rs#L232)?
it is not documented in the docstring above and I could not figure out what it should be.
The doctests appear to be just copies of the unit tests in this module.

pazz commented:
I am inclined to ignore the one complaint clippy has now that I made `pub(crate)` methods simply public:

```
error: struct `Ideal` has a public `len` method, but no `is_empty` method
   --> src/ideal.rs:103:5
    |
103 |     pub fn len(&self) -> usize {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#len_without_is_empty
    = note: `-D clippy::len-without-is-empty` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(clippy::len_without_is_empty)]`
```

as Ideals are by definition non-empty, I don't think adding an `is_empty` method is useful here, even if this would refer to the size of the representing data structure rather than the represented set.

We can either add this, rename that method `len`, tell clippy to shut up, by tagging it with

```rust
#[allow(clippy::len_without_is_empty)] 
```

I don't mind, your call.

Numero7 commented:
I believe "len" is improper and should be renamed as "dim"

Numero7 commented:
@pazz also "from_vec" should raise exception in case "vec" is empty and some comment should be added beside "pub struct Ideal(Vec<Coef>);" to specify a type invariant: the inner vector is never empty

pazz commented:
> I believe "len" is improper and should be renamed as "dim"



> I believe "len" is improper and should be renamed as "dim"

Got it; I renamed it to "dimension now" and added a docstring. 
Language servers for the win!
